gdal/port/cpl_vsil_s3.cpp

```diff
@@ -169,64 +284,51 @@ void VSICurlFilesystemHandler::AnalyseS3FileList(
                         osKey.resize(osKey.size()-1);
                     if( osKey.size() > osPrefix.size() )
                     {
-                        FileProp prop;
-                        prop.eExists = EXIST_YES;
-                        prop.bIsDirectory = true;
-                        prop.bHasComputedFileSize = true;
-                        prop.fileSize = 0;
-                        prop.mTime = 0;
+                        aoEntries.push_back(
+                            std::unique_ptr<VSIDIREntry>(new VSIDIREntry()));
+                        auto& entry = aoEntries.back();
+                        entry->pszName = CPLStrdup(osKey.c_str() + osPrefix.size());
+                        if( aoNameCount[entry->pszName] == 2 )
+                        {
+                            // Add a / suffix to disambiguish the situation
+                            // Normally we don't suffix directories with /, but
```

Comment:
I think this needs to be documented in VSIGetNextDirEntry(). 

Also trying to figure out why we have no alternative? — if the caller is expecting directories they need to call `entry.mod & S_IFDIR` to find out what it is anyway... in which case filenames don't need to be unique?

Reply:
Could be a bit confusing to have different objects identified by the same name

Reply:
Sure. Is also confusing to always have directories with no trailing `/` except in the case where there's a file of the same name. (I'd be ok always having a trailing `/` for directories but that's inconsistent with what's there already, I think that ship has sailed)

I think if it's documented in the API at least people may expect it.